### PR TITLE
chore(RHTAPWATCH-238): Alert rule for progressing argoCD apps

### PR DESCRIPTION
**Pull request by ShakedAviv50313**
* An alert rule for argoCD apps which were in "Progressing" state in the past 20 minutes, and are still not healthy.

### CODE REVIEW

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ernesgonzalez33) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:29]: Please replace "minuts" with "minutes"

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (Omeramsc) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:30]: ^ apply Ernesto's comment here too

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:25]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:24]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:52]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:53]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:72]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:73]: same

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:37]: Why should this alert? didn't it turn Healthy before the 20 min mark?

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:59]: if the comment is needed, try making it more readable. I first thought it is a leftover

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:57]: WDYM by flapping?

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [inline test/promql/tests/test_progressing_argocd.yaml:77]: why does it alert on the former test but not on this one?

**Review by Omeramsc — CHANGES_REQUESTED** [review]

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:29]: These are wrong summary and description, the application was in progress state for more than 50% of the previous 20 minutes.

File: test/promql/tests/test_progressing_argocd.yaml
Comment (avi-biton) [inline test/promql/tests/test_progressing_argocd.yaml:31]: As discussed, all the `0` should be updated to `_` (all over the tests)

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [reply test/promql/tests/test_progressing_argocd.yaml:37]: cool, I will however mention it in the rule.

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:30]: Empty line is missing at the EOF

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:22]: Since the values are not `1` or `0`, but `1` or `_`,  we saw that the use of `avg_over_time` is giving us wrong values.
The average will always be `1` even if it appears only once, in this series for example : `_x16 1 _ _`, we have 19 `_` and only one `1`, and the average is 1 since the `_` are not accounted as `0`

We have to come up with different approach

File: test/promql/tests/test_progressing_argocd.yaml
Comment (avi-biton) [reply test/promql/tests/test_progressing_argocd.yaml:77]: because this series:
```
- series: 'argocd_app_info{health_status="Progressing", name="fith-test-app", dest_namespace="Flapping-No-Alert"}'
        values: '1 0 0 1 0 0 1 1 1 0 0 0 0 1 0 1 1 1 1 0' 
```
it is `Progressing` 10 out of 20 times in the last 20 minutes, which is `=0.5`, and the rule is for `>0.5`

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:30]:
```suggestion
        description: "Application {{ $labels.name }} in namespace {{ $labels.dest_namespace }} was in 'Progressing' state in the last 20 minutes."

```

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: We have status healthy mentioned on the tests quite a lot, but the expression is no longer considering it. We should have the tests series mention other statuses which are not progressing or healthy.

File: test/promql/tests/test_progressing_argocd.yaml
Comment (yftacherzog) [inline test/promql/tests/test_progressing_argocd.yaml:1]: I'd also add a test case in which we have more than one alerting series.

File: test/promql/tests/test_progressing_argocd.yaml
Comment (Omeramsc) [reply test/promql/tests/test_progressing_argocd.yaml:77]: Should be mentioned in the comment, Let's not expect readers to count the ones and zeros...

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml:1]: Added:)

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: It is now changed, now "healthy" health status is mentioned only where it's still relevant

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml:48]: done:)

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: Done

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: It is no longer relevant since the rule is now changed and no longer consists of an average.

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: it changed, the rule has also changed so it's according to the new rule

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: Done

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: Done

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: It's changed, the rule now consists of a check whether it was "Progressing" in the last 20 minutes and also if it's not "Healthy" for the past 4 minutes.

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: In which cases is it relevant? The alert expression is not considering metrics with status=healthy.

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: We should instead have test cases that has metrics with labels which are not healthy or progressing in the last 4 minutes so we test the part of the expression that matches !=healthy.

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: Now the only comments with the word "Healthy" or it's variations are:
1.       # app was progessing and then Healthy for the last 5 min, should not alert
2.       # app was progessing and then Healthy for the last 3 min, should alert
3.       # the app is switching health states, alert should be triggered

which of them does not comply with these rules?

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: There's this line:
``` max_over_time(argocd_app_info{health_status!="Healthy", dest_namespace!~".+-tenant"}[4m]) == 1``` 
so in order for the tests to be clear and understandable there's a need to state whether the app is healthy and if so for how long, this is just in my opinion .

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: The thing is, the tests will pass just the same if you delete all series with status=healthy. I think that adding the healthy series in comment or to the first use case just to convey the idea might be helpful. I don't think we should have it on all test cases while the rule is ignoring it.

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]: ok, I removed it

**Review by Omeramsc — APPROVED** [review]
One minor comment, besides that LGTM

**Review by Omeramsc — APPROVED** [review]
One minor comment, besides that LGTM

File: test/promql/tests/test_progressing_argocd.yaml
Comment (ShakedAviv50313) [reply test/promql/tests/test_progressing_argocd.yaml]: Done

**Review by yftacherzog — APPROVED** [review]

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (avi-biton) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: Didn't we want to check if  `Progressing` > `Healthy` in the past 20 minutes?

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: @ShakedAviv50313 found a bug there. If there are no healthy readings inside the time frame, then the comparison will fail.

i.e. to have any operator between 2 series, we need each of them to have at least one reading within the time frame, otherwise the operation will not return any time series. This means that that option will not detect an app which is progressing throughout the time window (or flapping between progressing and anything other than healthy).

Comment (avi-biton) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: In this case we can use the `or vector(0)` which will return a vector of `0`s in case the query returns None or null:

```
(count_over_time(argocd_app_info{health_status="Progressing", dest_namespace!~".+-tenant"}[20m] or vector(0)) 
> ignoring (health_status) 
count_over_time(argocd_app_info{health_status="Healthy", dest_namespace!~".+-tenant"}[20m]) or vector(0)) 
and ignoring (health_status) 
max_over_time(argocd_app_info{health_status!="Healthy", dest_namespace!~".+-tenant"}[4m]) == 1
```

I haven't test it yet, so it might need a little tweaking, but I think that that's the approach

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: Indeed needs quite some testing but that might be a way forward here.

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: It passes all the tests!! so i'm changing this pr for the new rule, just for general knowledge i was looking at and trying to use the `absent(sample_query)` which return 1 if a query does not return anything, and 0 otherwise.

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: @avi-biton @yftacherzog can I please get a second review on this?

**Review by yftacherzog — APPROVED** [review]

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:24]:
```suggestion
        count_over_time(argocd_app_info{health_status="Healthy", dest_namespace!~".+-tenant"}[20m]) or vector(0))
```

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml]: done

**Review by avi-biton — APPROVED** [review]

**Pull request merged**
